### ISSUE TITLE
Add support for boolean mask indices inside jit

### ISSUE THREAD

**Issue by inversecrime** [google/jax#21994]
Hi,
basically, jax always complains when you use `.at[...]` with boolean masks inside of `jax.jit`, because boolean masks refer to a non-static number of elements.

However, in many cases, these operations are well-defined even in the context of static shapes.
Currently, it is possible to achieve this with `jnp.where` but not with `.at[...]`:
```python
import jax
import jax.numpy as jnp

def f(x):
    return x.at[x < 5].set(0)

def g(x):
    return jnp.where(x < 5, 0, x)

x = jnp.arange(10)
assert jnp.array_equal(f(x), g(x))  # type: ignore

print(jax.jit(g)(x))  # no error
print(jax.jit(f)(x))  # error
```

One could say "if this behavior is needed, just use `jnp.where`" - but with more complex indexing, `jnp.where` becomes less and less intuitive.
Consider the following example:
```python
def f(x):
    return x.at[:4, jnp.mod(x[0], 2) == 0].set(0)  # <-- far more intuitive (atleast for me)

def g(x):
    return x.at[:4].set(jnp.where(jnp.mod(x[0], 2) == 0, 0, x[:4]))  # <-- very abstract and unintuitive (atleast for me)

x = jnp.reshape(jnp.arange(100), shape=(10, 10))
assert jnp.array_equal(f(x), g(x))

print(jax.jit(g)(x))  # no error
print(jax.jit(f)(x))  # error
```

PS:
By the way, why is `type: ignore` needed in the first example code? The type checker complains here.

**Comment by jakevdp** [comment]
I've thought about this on several occasions, but the added flexibility has never seemed worth the added complexity in the code.

As @justinjfu mentioned, arr.at[...].*` is currently a way of spelling`lax.gather` and `lax.scatter`. These APIs do not support dynamic shapes, and so boolean indices cannot be used within JIT.

It's important to note that `arr.at[mask].set(arr)` is *inherently* a dynamically-shaped operation, because semantically `arr` must be broadcast-compatible with the entries of `arr` selected by `mask`. It's true that there is a special case when `arr` has size 1, because a size-1 axis can broadcast to any array shape: but the shape is still dynamic.

In the simplest case of a 1D array with a 1D mask, you could imagine forking the logic and lowering to `select` rather than `scatter`; but immediately someone would ask about `arr.at[0, mask]` or `arr.at[indices, mask]` or `arr.at[..., mask, :, 0, 1]`, or all the other variants of combined indices that it may be possible to lower to select (depending on the shape of the right-hand-side array). To get all of that logic correct would add a *lot* of complexity both to the implementation, and to the mental overhead of the user trying to figure out when an operation might be JIT-compatible.

By contrast, the current rule is very simple: boolean indices are not supported by JIT, and if you want to use `select` instead of `gather`/`scatter`, you can do so explicitly. In this case, I think simpler is better.

**Comment by inversecrime** [comment]
Thanks for the answers. Just one more question - how does `.at` work with boolean indices when not compiled?

**Comment by jakevdp** [comment]
> how does `.at` work with boolean indices when not compiled?

It converts them to integer indices, then passes those integer indices to `scatter` or `gather`.

**Comment by inversecrime** [comment]
ooh I see. Well, would still be nice to have at some point...
